packages/core/src/anchor/ethereum/ethereum-anchor-service.ts

```diff
@@ -1,116 +1,68 @@
-  timer,
-  of,
-  defer,
-  expand,
-  lastValueFrom,
-  Subject,
-  type OperatorFunction,
-  type MonoTypeOperatorFunction,
-} from 'rxjs'
-import { concatMap, catchError, map, retry, tap } from 'rxjs/operators'
-import { CAR } from 'cartonne'
+import { Subject, type Observable } from 'rxjs'
+import type { CAR } from 'cartonne'
 import { AnchorRequestCarFileReader } from '../anchor-request-car-file-reader.js'
-import { CASResponseOrError, ErrorResponse, AnchorRequestStatusName } from '@ceramicnetwork/codecs'
-import { decode } from 'codeco'
+import { AnchorRequestStatusName } from '@ceramicnetwork/codecs'
 import { EthereumAnchorValidator } from './ethereum-anchor-validator.js'
 import type {
   AnchorService,
   AnchorServiceAuth,
   AnchorValidator,
   AuthenticatedAnchorService,
+  CASClient,
 } from '../anchor-service.js'
-import { CasConnectionError, MaxAnchorPollingError } from '../anchor-service.js'
+import type { AnchorRequestStore } from '../../store/anchor-request-store.js'
+import { NotSingleChainError, type AnchorLoopHandler } from '../anchor-service.js'
+import { AnchorProcessingLoop } from '../anchor-processing-loop.js'
 import { RemoteCAS } from './remote-cas.js'
+import { doNotWait } from '../../ancillary/do-not-wait.js'
+import { NamedTaskQueue } from '../../state-management/named-task-queue.js'
 
 const DEFAULT_POLL_INTERVAL = 60_000 // 60 seconds
-const MAX_POLL_TIME = 86_400_000 // 24 hours
-
-/**
- * Parse JSON that CAS returns.
- */
-function parseResponse(streamId: StreamID, tip: CID, json: unknown): AnchorEvent {
-  const parsed = decode(CASResponseOrError, json)
-  if (ErrorResponse.is(parsed)) {
-    return {
-      status: AnchorRequestStatusName.FAILED,
-      streamId: streamId,
-      cid: tip,
-      message: parsed.error,
-    }
-  } else {
-    if (parsed.status === AnchorRequestStatusName.COMPLETED) {
-      return {
-        status: parsed.status,
-        streamId: parsed.streamId,
-        cid: parsed.cid,
-        message: parsed.message,
-        witnessCar: parsed.witnessCar,
-      }
-    }
-    return {
-      status: parsed.status,
-      streamId: parsed.streamId,
-      cid: parsed.cid,
-      message: parsed.message,
-    }
-  }
-}
-
-function announcePending(streamId: StreamID, tip: CID): Observable<AnchorEvent> {
-  return of({
-    status: AnchorRequestStatusName.PENDING,
-    streamId: streamId,
-    cid: tip,
-    message: 'Sending anchoring request',
-  })
-}
+const BATCH_SIZE = 10
 
 /**
  * Ethereum anchor service that stores root CIDs on Ethereum blockchain
  */
 export class EthereumAnchorService implements AnchorService {
   readonly #logger: DiagnosticsLogger
-  /**
-   * Retry a request to CAS every +pollInterval+ milliseconds.
-   */
-  readonly #pollInterval: number
-  readonly #maxPollTime: number
+  #loop: AnchorProcessingLoop
+  readonly #enableAnchorPollingLoop: boolean
   readonly #events: Subject<AnchorEvent>
-  readonly #cas: RemoteCAS
+  readonly #queue: NamedTaskQueue
```

Comment:
Can we name this `#anchorStoreQueue` or something like that, and add comments explaining why we need it?